File: trunk/source/math/BoundingSphere.cpp
#include "stdafx.h"
/*
* Copyright (c) 2007-2011 SlimDX Group
* 
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
* 
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
* 
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*/

#include <d3dx9.h>

#include "../SlimDXException.h"

#include "BoundingSphere.h"
#include "BoundingBox.h"
#include "Ray.h"
#include "Plane.h"

using namespace System;
using namespace System::Globalization;

namespace SlimDX
{
	BoundingSphere::BoundingSphere( Vector3 center, float radius )
	{
		Center = center;
		Radius = radius;
	}

	ContainmentType BoundingSphere::Contains( BoundingSphere sphere, BoundingBox box )
	{
		Vector3 vector;

		if( !BoundingBox::Intersects( box, sphere ) )
			return ContainmentType::Disjoint;

		float radius = sphere.Radius * sphere.Radius;
		vector.X = sphere.Center.X - box.Minimum.X;
		vector.Y = sphere.Center.Y - box.Maximum.Y;
		vector.Z = sphere.Center.Z - box.Maximum.Z;

		if( vector.LengthSquared() > radius )
			return ContainmentType::Intersects;

		vector.X = sphere.Center.X - box.Maximum.X;
		vector.Y = sphere.Center.Y - box.Maximum.Y;
		vector.Z = sphere.Center.Z - box.Maximum.Z;

		if( vector.LengthSquared() > radius )
			return ContainmentType::Intersects;

		vector.X = sphere.Center.X - box.Maximum.X;
		vector.Y = sphere.Center.Y - box.Minimum.Y;
		vector.Z = sphere.Center.Z - box.Maximum.Z;

		if( vector.LengthSquared() > radius )
			return ContainmentType::Intersects;

		vector.X = sphere.Center.X - box.Minimum.X;
		vector.Y = sphere.Center.Y - box.Minimum.Y;
		vector.Z = sphere.Center.Z - box.Maximum.Z;

		if( vector.LengthSquared() > radius )
			return ContainmentType::Intersects;

		vector.X = sphere.Center.X - box.Minimum.X;
		vector.Y = sphere.Center.Y - box.Maximum.Y;
		vector.Z = sphere.Center.Z - box.Minimum.Z;

		if( vector.LengthSquared() > radius )
			return ContainmentType::Intersects;

		vector.X = sphere.Center.X - box.Maximum.X;
		vector.Y = sphere.Center.Y - box.Maximum.Y;
		vector.Z = sphere.Center.Z - box.Minimum.Z;

		if( vector.LengthSquared() > radius )
			return ContainmentType::Intersects;

		vector.X = sphere.Center.X - box.Maximum.X;
		vector.Y = sphere.Center.Y - box.Minimum.Y;
		vector.Z = sphere.Center.Z - box.Minimum.Z;

		if( vector.LengthSquared() > radius )
			return ContainmentType::Intersects;

		vector.X = sphere.Center.X - box.Minimum.X;
		vector.Y = sphere.Center.Y - box.Minimum.Y;
		vector.Z = sphere.Center.Z - box.Minimum.Z;

		if( vector.LengthSquared() > radius )
			return ContainmentType::Intersects;

		return ContainmentType::Contains;
	}

	ContainmentType BoundingSphere::Contains( BoundingSphere sphere1, BoundingSphere sphere2 )
	{
		float distance;
		float x = sphere1.Center.X - sphere2.Center.X;
		float y = sphere1.Center.Y - sphere2.Center.Y;
		float z = sphere1.Center.Z - sphere2.Center.Z;

		distance = static_cast<float>( Math::Sqrt( (x * x) + (y * y) + (z * z) ) );
		float radius = sphere1.Radius;
		float radius2 = sphere2.Radius;

		if( radius + radius2 < distance )
			return ContainmentType::Disjoint;

		if( radius - radius2 < distance )
			return ContainmentType::Intersects;

		return ContainmentType::Contains;
	}

	ContainmentType BoundingSphere::Contains( BoundingSphere sphere, Vector3 vector )
	{
		float x = vector.X - sphere.Center.X;
		float y = vector.Y - sphere.Center.Y;
		float z = vector.Z - sphere.Center.Z;

		float distance = (x * x) + (y * y) + (z * z);

		if( distance >= (sphere.Radius * sphere.Radius) )
			return ContainmentType::Disjoint;

		return ContainmentType::Contains;
	}

	BoundingSphere BoundingSphere::FromBox( BoundingBox box )
	{
		BoundingSphere sphere;
		Vector3::Lerp( box.Minimum, box.Maximum, 0.5f, sphere.Center );

		float x = box.Minimum.X - box.Maximum.X;
		float y = box.Minimum.Y - box.Maximum.Y;
		float z = box.Minimum.Z - box.Maximum.Z;

		float distance = static_cast<float>( Math::Sqrt( (x * x) + (y * y) + (z * z) ) );

		sphere.Radius = distance * 0.5f;

		return sphere;
	}

	BoundingSphere BoundingSphere::FromPoints( array<Vector3>^ points )
	{
		D3DXVECTOR3 center;
		FLOAT radius;
		pin_ptr<Vector3> pinnedPoints = &points[0];

		HRESULT hr = D3DXComputeBoundingSphere( reinterpret_cast<const D3DXVECTOR3*>( pinnedPoints ), points->Length, sizeof(float) * 3, &center, &radius );
		if( RECORD_SDX( hr ).IsFailure )
			return BoundingSphere();

		BoundingSphere sphere;
		sphere.Center = Vector3( center.x, center.y, center.z );
		sphere.Radius = radius;

		return sphere;
	}

	BoundingSphere BoundingSphere::Merge( BoundingSphere sphere1, BoundingSphere sphere2 )
	{
		BoundingSphere sphere;
		Vector3 difference = sphere2.Center - sphere1.Center;

		float length = difference.Length();
		float radius = sphere1.Radius;
		float radius2 = sphere2.Radius;

		if( radius + radius2 >= length)
		{
			if( radius - radius2 >= length )
				return sphere1;

			if( radius2 - radius >= length )
				return sphere2;
		}

		Vector3 vector = difference * ( 1.0f / length );
		float min = Math::Min( -radius, length - radius2 );
		float max = ( Math::Max( radius, length + radius2 ) - min ) * 0.5f;

		sphere.Center = sphere1.Center + vector * ( max + min );
		sphere.Radius = max;

		return sphere;
	}

	bool BoundingSphere::Intersects( BoundingSphere sphere, BoundingBox box )
	{
		return BoundingBox::Intersects( box, sphere );
	}

	bool BoundingSphere::Intersects( BoundingSphere sphere1, BoundingSphere sphere2 )
	{
		float distance;
		distance = Vector3::DistanceSquared( sphere1.Center, sphere2.Center );
		float radius = sphere1.Radius;
		float radius2 = sphere2.Radius;

		if( (radius * radius) + (2.0f * radius * radius2) + (radius2 * radius2) <= distance )
			return false;

		return true;
	}

	bool BoundingSphere::Intersects( BoundingSphere sphere, Ray ray, [Out] float% distance )
	{
		return Ray::Intersects( ray, sphere, distance );
	}

	PlaneIntersectionType BoundingSphere::Intersects( BoundingSphere sphere, Plane plane )
	{
		return Plane::Intersects( plane, sphere );
	}

	bool BoundingSphere::operator == ( BoundingSphere left, BoundingSphere right )
	{
		return BoundingSphere::Equals( left, right );
	}

	bool BoundingSphere::operator != ( BoundingSphere left, BoundingSphere right )
	{
		return !BoundingSphere::Equals( left, right );
	}

	String^ BoundingSphere::ToString()
	{
		return String::Format( CultureInfo::CurrentCulture, "Center:{0} Radius:{1}", Center.ToString(), Radius.ToString(CultureInfo::CurrentCulture) );
	}

	int BoundingSphere::GetHashCode()
	{
		return Center.GetHashCode() + Radius.GetHashCode();
	}

	bool BoundingSphere::Equals( Object^ value )
	{
		if( value == nullptr )
			return false;

		if( value->GetType() != GetType() )
			return false;

		return Equals( safe_cast<BoundingSphere>( value ) );
	}

	bool BoundingSphere::Equals( BoundingSphere value )
	{
		return ( Center == value.Center && Radius == value.Radius );
	}

	bool BoundingSphere::Equals( BoundingSphere% value1, BoundingSphere% value2 )
	{
		return ( value1.Center == value2.Center && value1.Radius == value2.Radius );
	}
}

File: SlimMath/BoundingSphere.cs
/*
* Copyright (c) 2007-2010 SlimDX Group
* 
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
* 
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
* 
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*/

using System;
using System.Globalization;
using System.Runtime.InteropServices;
using System.ComponentModel;

namespace SlimMath
{
    /// <summary>
    /// Represents a bounding sphere in three dimensional space.
    /// </summary>
    [Serializable]
    [StructLayout(LayoutKind.Sequential, Pack = 4)]
    public struct BoundingSphere : IEquatable<BoundingSphere>, IFormattable
    {
        /// <summary>
        /// The center of the sphere in three dimensional space.
        /// </summary>
        public Vector3 Center;

        /// <summary>
        /// The radious of the sphere.
        /// </summary>
        public float Radius;

        /// <summary>
        /// Initializes a new instance of the <see cref="SlimMath.BoundingBox"/> struct.
        /// </summary>
        /// <param name="center">The center of the sphere.</param>
        /// <param name="radius">The radius of the sphere.</param>
        public BoundingSphere(Vector3 center, float radius)
        {
            this.Center = center;
            this.Radius = radius;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="SlimMath.BoundingBox"/> struct.
        /// </summary>
        /// <param name="centerX">The x-coordinate for the center of the sphere.</param>
        /// <param name="centerY">The y-coordinate for the center of the sphere.</param>
        /// <param name="centerZ">The z-coordinate for the center of the sphere.</param>
        /// <param name="radius">The radius of the sphere.</param>
        public BoundingSphere(float centerX, float centerY, float centerZ, float radius)
        {
            this.Center = new Vector3(centerX, centerY, centerZ);
            this.Radius = radius;
        }

        /// <summary>
        /// Determines if there is an intersection between the current object and a <see cref="SlimMath.Ray"/>.
        /// </summary>
        /// <param name="ray">The ray to test.</param>
        /// <returns>Whether the two objects intersected.</returns>
        public bool Intersects(ref Ray ray)
        {
            float distance;
            return Collision.RayIntersectsSphere(ref ray, ref this, out distance);
        }

        /// <summary>
        /// Determines if there is an intersection between the current object and a <see cref="SlimMath.Ray"/>.
        /// </summary>
        /// <param name="ray">The ray to test.</param>
        /// <param name="distance">When the method completes, contains the distance of the intersection,
        /// or 0 if there was no intersection.</param>
        /// <returns>Whether the two objects intersected.</returns>
        public bool Intersects(ref Ray ray, out float distance)
        {
            return Collision.RayIntersectsSphere(ref ray, ref this, out distance);
        }

        /// <summary>
        /// Determines if there is an intersection between the current object and a <see cref="SlimMath.Ray"/>.
        /// </summary>
        /// <param name="ray">The ray to test.</param>
        /// <param name="point">When the method completes, contains the point of intersection,
        /// or <see cref="SlimMath.Vector3.Zero"/> if there was no intersection.</param>
        /// <returns>Whether the two objects intersected.</returns>
        public bool Intersects(ref Ray ray, out Vector3 point)
        {
            return Collision.RayIntersectsSphere(ref ray, ref this, out point);
        }

        /// <summary>
        /// Determines if there is an intersection between the current object and a <see cref="SlimMath.Plane"/>.
        /// </summary>
        /// <param name="plane">The plane to test.</param>
        /// <returns>Whether the two objects intersected.</returns>
        public PlaneIntersectionType Intersects(ref Plane plane)
        {
            return Collision.PlaneIntersectsSphere(ref plane, ref this);
        }

        /// <summary>
        /// Determines if there is an intersection between the current object and a triangle.
        /// </summary>
        /// <param name="vertex1">The first vertex of the triangle to test.</param>
        /// <param name="vertex2">The second vertex of the triagnle to test.</param>
        /// <param name="vertex3">The third vertex of the triangle to test.</param>
        /// <returns>Whether the two objects intersected.</returns>
        public bool Intersects(ref Vector3 vertex1, ref Vector3 vertex2, ref Vector3 vertex3)
        {
            return Collision.SphereIntersectsTriangle(ref this, ref vertex1, ref vertex2, ref vertex3);
        }

        /// <summary>
        /// Determines if there is an intersection between the current object and a <see cref="SlimMath.BoundingBox"/>.
        /// </summary>
        /// <param name="box">The box to test.</param>
        /// <returns>Whether the two objects intersected.</returns>
        public bool Intersects(ref BoundingBox box)
        {
            return Collision.BoxIntersectsSphere(ref box, ref this);
        }

        /// <summary>
        /// Determines if there is an intersection between the current object and a <see cref="SlimMath.BoundingSphere"/>.
        /// </summary>
        /// <param name="sphere">The sphere to test.</param>
        /// <returns>Whether the two objects intersected.</returns>
        public bool Intersects(ref BoundingSphere sphere)
        {
            return Collision.SphereIntersectsSphere(ref this, ref sphere);
        }

        /// <summary>
        /// Determines whether the current objects contains a point.
        /// </summary>
        /// <param name="point">The point to test.</param>
        /// <returns>The type of containment the two objects have.</returns>
        public ContainmentType Contains(ref Vector3 point)
        {
            return Collision.SphereContainsPoint(ref this, ref point);
        }

        /// <summary>
        /// Determines whether the current objects contains a triangle.
        /// </summary>
        /// <param name="vertex1">The first vertex of the triangle to test.</param>
        /// <param name="vertex2">The second vertex of the triagnle to test.</param>
        /// <param name="vertex3">The third vertex of the triangle to test.</param>
        /// <returns>The type of containment the two objects have.</returns>
        public ContainmentType Contains(ref Vector3 vertex1, ref Vector3 vertex2, ref Vector3 vertex3)
        {
            return Collision.SphereContainsTriangle(ref this, ref vertex1, ref vertex2, ref vertex3);
        }

        /// <summary>
        /// Determines whether the current objects contains a <see cref="SlimMath.BoundingBox"/>.
        /// </summary>
        /// <param name="box">The box to test.</param>
        /// <returns>The type of containment the two objects have.</returns>
        public ContainmentType Contains(ref BoundingBox box)
        {
            return Collision.SphereContainsBox(ref this, ref box);
        }

        /// <summary>
        /// Determines whether the current objects contains a <see cref="SlimMath.BoundingSphere"/>.
        /// </summary>
        /// <param name="sphere">The sphere to test.</param>
        /// <returns>The type of containment the two objects have.</returns>
        public ContainmentType Contains(ref BoundingSphere sphere)
        {
            return Collision.SphereContainsSphere(ref this, ref sphere);
        }

        /// <summary>
        /// Translates and scales this instance by a <see cref="SlimMath.Matrix"/>.
        /// </summary>
        /// <param name="matrix">The <see cref="SlimMath.Matrix"/> to transform this instance by.</param>
        /// <param name="result">When the method completes, contains the transformed <see cref="SlimMath.BoundingSphere"/>.</param>
        /// <remarks>
        /// The result of this operation is undefined if the matrix contains any transformations other than
        /// translation, rotation, and uniform scaling.
        /// </remarks>
        public void Transform(ref Matrix matrix, out BoundingSphere result)
        {
            Vector3.TransformCoordinate(ref Center, ref matrix, out result.Center);

            float row1 = (matrix.M11 * matrix.M11) + (matrix.M12 * matrix.M12) + (matrix.M13 * matrix.M13);
            float row2 = (matrix.M21 * matrix.M21) + (matrix.M22 * matrix.M22) + (matrix.M23 * matrix.M23);
            float row3 = (matrix.M31 * matrix.M31) + (matrix.M32 * matrix.M32) + (matrix.M33 * matrix.M33);
            float num = Math.Max(row1, Math.Max(row2, row3));

            result.Radius = Radius * (float)Math.Sqrt(num);

        }

        /// <summary>
        /// Translates and scales this instance by a <see cref="SlimMath.Matrix"/>.
        /// </summary>
        /// <param name="matrix">The <see cref="SlimMath.Matrix"/> to transform this instance by.</param>
        /// <returns>The transformed <see cref="SlimMath.BoundingSphere"/>.</returns>
        /// <remarks>
        /// The result of this operation is undefined if the matrix contains any transformations other than
        /// translation, rotation, and uniform scaling.
        /// </remarks>
        public BoundingSphere Transform(Matrix matrix)
        {
            BoundingSphere result;
            Transform(ref matrix, out result);
            return result;
        }

        /// <summary>
        /// Generates a supporting point for this instance.
        /// </summary>
        /// <param name="direction">The direction for which to build the supporting point.</param>
        /// <param name="result">When the method completes, contains the supporting point.</param>
        public void SupportMapping(ref Vector3 direction, out Vector3 result)
        {
            Collision.SupportPoint(ref this, ref direction, out result);
        }

        /// <summary>
        /// Generates a support mapping for this instance.
        /// </summary>
        /// <param name="direction">The direction for which to build the support mapping.</param>
        /// <returns>The resulting support mapping.</returns>
        public Vector3 SupportMapping(Vector3 direction)
        {
            Vector3 result;
            SupportMapping(ref direction, out result);
            return result;
        }

        /// <summary>
        /// Constructs a <see cref="SlimMath.BoundingSphere"/> that fully contains the given points.
        /// </summary>
        /// <param name="points">The points that will be contained by the sphere.</param>
        /// <param name="result">When the method completes, contains the newly constructed bounding sphere.</param>
        public static void FromPoints(Vector3[] points, out BoundingSphere result)
        {
            //Find the center of all points.
            Vector3 center = Vector3.Zero;
            for (int i = 0; i < points.Length; ++i)
            {
                Vector3.Add(ref points[i], ref center, out center);
            }

            //This is the center of our sphere.
            center /= (float)points.Length;

            //Find the radius of the sphere
            float radius = 0f;
            for (int i = 0; i < points.Length; ++i)
            {
                //We are doing a relative distance comparison to find the maximum distance
                //from the center of our sphere.
                float distance;
                Vector3.DistanceSquared(ref center, ref points[i], out distance);

                if (distance > radius)
                    radius = distance;
            }

            //Find the real distance from the DistanceSquared.
            radius = (float)Math.Sqrt(radius);

            //Construct the sphere.
            result.Center = center;
            result.Radius = radius;
        }

        /// <summary>
        /// Constructs a <see cref="SlimMath.BoundingSphere"/> that fully contains the given points.
        /// </summary>
        /// <param name="points">The points that will be contained by the sphere.</param>
        /// <returns>The newly constructed bounding sphere.</returns>
        public static BoundingSphere FromPoints(Vector3[] points)
        {
            BoundingSphere result;
            FromPoints(points, out result);
            return result;
        }

        /// <summary>
        /// Constructs a <see cref="SlimMath.BoundingSphere"/> from a given box.
        /// </summary>
        /// <param name="box">The box that will designate the extents of the sphere.</param>
        /// <param name="result">When the method completes, the newly constructed bounding sphere.</param>
        public static void FromBox(ref BoundingBox box, out BoundingSphere result)
        {
            Vector3.Lerp(ref box.Minimum, ref box.Maximum, 0.5f, out result.Center);

            float x = box.Minimum.X - box.Maximum.X;
            float y = box.Minimum.Y - box.Maximum.Y;
            float z = box.Minimum.Z - box.Maximum.Z;

            float distance = (float)(Math.Sqrt((x * x) + (y * y) + (z * z)));
            result.Radius = distance * 0.5f;
        }

        /// <summary>
        /// Constructs a <see cref="SlimMath.BoundingSphere"/> from a given box.
        /// </summary>
        /// <param name="box">The box that will designate the extents of the sphere.</param>
        /// <returns>The newly constructed bounding sphere.</returns>
        public static BoundingSphere FromBox(BoundingBox box)
        {
            BoundingSphere result;
            FromBox(ref box, out result);
            return result;
        }

        /// <summary>
        /// Constructs a <see cref="SlimMath.BoundingSphere"/> that is the as large as the total combined area of the two specified spheres.
        /// </summary>
        /// <param name="value1">The first sphere to merge.</param>
        /// <param name="value2">The second sphere to merge.</param>
        /// <param name="result">When the method completes, contains the newly constructed bounding sphere.</param>
        public static void Merge(ref BoundingSphere value1, ref BoundingSphere value2, out BoundingSphere result)
        {
            Vector3 difference = value2.Center - value1.Center;

            float length = difference.Length;
            float radius = value1.Radius;
            float radius2 = value2.Radius;

            if (radius + radius2 >= length)
            {
                if (radius - radius2 >= length)
                {
                    result = value1;
                    return;
                }

                if (radius2 - radius >= length)
                {
                    result = value2;
                    return;
                }
            }

            Vector3 vector = difference * (1.0f / length);
            float min = Math.Min(-radius, length - radius2);
            float max = (Math.Max(radius, length + radius2) - min) * 0.5f;

            result.Center = value1.Center + vector * (max + min);
            result.Radius = max;
        }

        /// <summary>
        /// Constructs a <see cref="SlimMath.BoundingSphere"/> that is the as large as the total combined area of the two specified spheres.
        /// </summary>
        /// <param name="value1">The first sphere to merge.</param>
        /// <param name="value2">The second sphere to merge.</param>
        /// <returns>The newly constructed bounding sphere.</returns>
        public static BoundingSphere Merge(BoundingSphere value1, BoundingSphere value2)
        {
            BoundingSphere result;
            Merge(ref value1, ref value2, out result);
            return result;
        }

        /// <summary>
        /// Tests for equality between two objects.
        /// </summary>
        /// <param name="left">The first value to compare.</param>
        /// <param name="right">The second value to compare.</param>
        /// <returns><c>true</c> if <paramref name="left"/> has the same value as <paramref name="right"/>; otherwise, <c>false</c>.</returns>
        public static bool operator ==(BoundingSphere left, BoundingSphere right)
        {
            return left.Equals(right);
        }

        /// <summary>
        /// Tests for inequality between two objects.
        /// </summary>
        /// <param name="left">The first value to compare.</param>
        /// <param name="right">The second value to compare.</param>
        /// <returns><c>true</c> if <paramref name="left"/> has a different value than <paramref name="right"/>; otherwise, <c>false</c>.</returns>
        public static bool operator !=(BoundingSphere left, BoundingSphere right)
        {
            return !left.Equals(right);
        }

        /// <summary>
        /// Returns a <see cref="System.String"/> that represents this instance.
        /// </summary>
        /// <returns>
        /// A <see cref="System.String"/> that represents this instance.
        /// </returns>
        public override string ToString()
        {
            return string.Format(CultureInfo.CurrentCulture, "Center:{0} Radius:{1}", Center.ToString(), Radius.ToString());
        }

        /// <summary>
        /// Returns a <see cref="System.String"/> that represents this instance.
        /// </summary>
        /// <param name="format">The format.</param>
        /// <returns>
        /// A <see cref="System.String"/> that represents this instance.
        /// </returns>
        public string ToString(string format)
        {
            if (format == null)
                return ToString();

            return string.Format(CultureInfo.CurrentCulture, "Center:{0} Radius:{1}", Center.ToString(format, CultureInfo.CurrentCulture),
                Radius.ToString(format, CultureInfo.CurrentCulture));
        }

        /// <summary>
        /// Returns a <see cref="System.String"/> that represents this instance.
        /// </summary>
        /// <param name="formatProvider">The format provider.</param>
        /// <returns>
        /// A <see cref="System.String"/> that represents this instance.
        /// </returns>
        public string ToString(IFormatProvider formatProvider)
        {
            return string.Format(formatProvider, "Center:{0} Radius:{1}", Center.ToString(), Radius.ToString());
        }

        /// <summary>
        /// Returns a <see cref="System.String"/> that represents this instance.
        /// </summary>
        /// <param name="format">The format.</param>
        /// <param name="formatProvider">The format provider.</param>
        /// <returns>
        /// A <see cref="System.String"/> that represents this instance.
        /// </returns>
        public string ToString(string format, IFormatProvider formatProvider)
        {
            if (format == null)
                return ToString(formatProvider);

            return string.Format(formatProvider, "Center:{0} Radius:{1}", Center.ToString(format, formatProvider),
                Radius.ToString(format, formatProvider));
        }

        /// <summary>
        /// Returns a hash code for this instance.
        /// </summary>
        /// <returns>
        /// A hash code for this instance, suitable for use in hashing algorithms and data structures like a hash table. 
        /// </returns>
        public override int GetHashCode()
        {
            return Center.GetHashCode() + Radius.GetHashCode();
        }

        /// <summary>
        /// Determines whether the specified <see cref="SlimMath.Vector4"/> is equal to this instance.
        /// </summary>
        /// <param name="value">The <see cref="SlimMath.Vector4"/> to compare with this instance.</param>
        /// <returns>
        /// <c>true</c> if the specified <see cref="SlimMath.Vector4"/> is equal to this instance; otherwise, <c>false</c>.
        /// </returns>
        public bool Equals(BoundingSphere value)
        {
            return Center == value.Center && Radius == value.Radius;
        }

        /// <summary>
        /// Determines whether the specified <see cref="System.Object"/> is equal to this instance.
        /// </summary>
        /// <param name="obj">The <see cref="System.Object"/> to compare with this instance.</param>
        /// <returns>
        /// <c>true</c> if the specified <see cref="System.Object"/> is equal to this instance; otherwise, <c>false</c>.
        /// </returns>
        public override bool Equals(object obj)
        {
            if (obj == null)
                return false;

            if (obj.GetType() != GetType())
                return false;

            return Equals((BoundingSphere)obj);
        }

#if SlimDX1xInterop
        /// <summary>
        /// Performs an implicit conversion from <see cref="SlimMath.BoundingSphere"/> to <see cref="SlimDX.BoundingSphere"/>.
        /// </summary>
        /// <param name="value">The value.</param>
        /// <returns>The result of the conversion.</returns>
        public static implicit operator SlimDX.BoundingSphere(BoundingSphere value)
        {
            return new SlimDX.BoundingSphere(value.Center, value.Radius);
        }

        /// <summary>
        /// Performs an implicit conversion from <see cref="SlimDX.BoundingSphere"/> to <see cref="SlimMath.BoundingSphere"/>.
        /// </summary>
        /// <param name="value">The value.</param>
        /// <returns>The result of the conversion.</returns>
        public static implicit operator BoundingSphere(SlimDX.BoundingSphere value)
        {
            return new BoundingSphere(value.Center, value.Radius);
        }
#endif

#if XnaInterop
        /// <summary>
        /// Performs an implicit conversion from <see cref="SlimMath.BoundingSphere"/> to <see cref="Microsoft.Xna.Framework.BoundingSphere"/>.
        /// </summary>
        /// <param name="value">The value.</param>
        /// <returns>The result of the conversion.</returns>
        public static implicit operator Microsoft.Xna.Framework.BoundingSphere(BoundingSphere value)
        {
            return new Microsoft.Xna.Framework.BoundingSphere(value.Center, value.Radius);
        }

        /// <summary>
        /// Performs an implicit conversion from <see cref="Microsoft.Xna.Framework.BoundingSphere"/> to <see cref="SlimMath.BoundingSphere"/>.
        /// </summary>
        /// <param name="value">The value.</param>
        /// <returns>The result of the conversion.</returns>
        public static implicit operator BoundingSphere(Microsoft.Xna.Framework.BoundingSphere value)
        {
            return new BoundingSphere(value.Center, value.Radius);
        }
#endif
    }
}
